spec/rubocop/cop/thread_safety/rack_middleware_instance_variable_spec.rb

```diff
@@ -0,0 +1,410 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::ThreadSafety::RackMiddlewareInstanceVariable, :config do
+  let(:msg) { 'Avoid instance variables in Rack middleware.' }
+
+  context 'with unrelated source' do
+    it { expect_no_offenses '' }
+
+    specify do
+      expect_no_offenses(<<~RUBY)
+        class SomeClass
+          def initialize(user)
+            @user = user
+          end
+        end
+      RUBY
+    end
+
+    specify do
+      expect_no_offenses(<<~RUBY)
+        class SomeClass
+          def initialize(user, context)
+            @user = user
+            @context = context
+          end
+        end
+      RUBY
+    end
+
+    specify do
+      expect_no_offenses(<<~RUBY)
+        class SomeClass
+          def call(env)
+            @env = env
+          end
+        end
+      RUBY
+    end
+
+    specify do
+      expect_no_offenses(<<~RUBY)
+        class SomeClass
+          def initialize(user, context)
+            @user = user
+            @context = context
+          end
+
+          def call
+            [@user, @context]
+          end
+        end
+      RUBY
+    end
+
+    specify do
+      expect_no_offenses(<<~RUBY)
+        class SomeClass
+          def initialize(app)
+            @user = User.new
+          end
+
+          def call(env)
+            @x = TOPLEVEL_BINDING
+          end
+        end
+      RUBY
+    end
+
+    specify do
+      expect_no_offenses(<<~RUBY)
+        class SomeClass
+          def initialize(app)
+            @app = app
+            @user = User.new
+          end
+
+          def call(env, user)
+            @x = TOPLEVEL_BINDING
+          end
+        end
+      RUBY
+    end
+  end
+
+  it 'does not register an offense' do
+    expect_no_offenses(<<~RUBY)
+      class TestMiddleware
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          @app.call(env)
+        end
+      end
+    RUBY
+  end
+
+  it 'registers an offense' do
+    expect_offense(<<~RUBY)
+      class TestMiddleware
+        def initialize(app)
+          @app = app
+          @foo = 1
+          ^^^^^^^^ #{msg}
```

Comment:
> I would also think that reading their values in call is safe

The problem here is that we can't know for sure if reading instance variable is followed by mutation. Nobody does this:

```ruby
class IdentityMiddleware
  def initialize(app, policies:)
    @app = app
    @policies = policies
  end

  def call(env)
    @policies # void expession, safe
    @policies != [] # another void expression, also safe

    @app.call(env)
  end
end
```

instead they do this:

```ruby
class IdentityMiddleware
  def initialize(app, policies:)
    @app = app
    @policies = policies
  end

  def call(env)
    # no way to tell of this is safe
    if PolicyValidator.acceptable? @policies, fetch_applicable_policy_from_env(env)
      @app.call(env)
    else
      [418, {}, []}
    end
  end
end
```

which corresponds to this sexp:

```shell
ruby-parse --34 --legacy -e "$(cat <<EOL
class IdentityMiddleware
  def initialize(app, policies:)
    @app = app
    @policies = policies
  end

  def call(env)
    if PolicyValidator.acceptable? @policies, fetch_applicable_policy_from_env(env)
      @app.call(env)
    else
      [418, {}, []]
    end
  end
end
EOL
)"

<< SKIPPED >>
(send
  (const nil :PolicyValidator) :acceptable?
    (ivar :@policies)
    (send nil :fetch_applicable_policy_from_env
    lvar :env)))
<< SKIPPED >>
```

Note the `(ivar :@policies)` here. In other words, we can't know for sure if `acceptable?` mutates the argument or not; we also don't know if `@policies` is mutable or not.

> I would think that setting other ivars in initialize is safe.

That's true, but this trick allows us not to track `attr_reader` in class definition (for instance, I prefer it in favor of addressing the ivar directly). I think it's possible to remove these offenses by tracking `attr_reader`, let me know if you'd like it to be this way.

In general, as always with rubocop that's all about choosing between "no missed offenses, some false positives" and "some missed offenses, almost no false positives". In case or rack middleware, people do not usually have a lot of them in their projects, so FPs do not hurt a lot.

Reply:
Thank you for the thorough explanation.